### ISSUE TITLE
Next wave of integration from Vale & HACL*/dev

### ISSUE THREAD

**Issue by msprotz** [project-everest/hacl-star#145]
:warning::warning::warning:
always close the page then reopen it before editing this issue
:warning::warning::warning:

Hello,

After a somewhat lengthy discussion and a review of all the moving pieces, here is an updated list of tasks and sub-tasks along with tentative assignments. The logistics of this integration effort are as follows:
- all developments to be merged into the branch fstar-master (which posts notifications to #hacl-build); please branch off, get a green, then merge back into this branch -- ask Jonathan for help if CI is acting up
- ~~we are aiming for an alpha release with only inevitable assumes leftover **for April 2nd**; the short deadline is due to an upcoming blog post & magazine article.~~

First, a list of completed sub-tasks (thanks to all who participated already!).

- [x] `--admit_smt_queries true` in `lib/Lib.ByteBuffer.fst` (line 77 onwards)
- [x] `fill_blocks` regression in `Lib.Buffer.fst` (Santiago)
- [x] use right bundling arguments (Jonathan)
- [x] drop old poly1305 (Jonathan)
- [x] two admit() in FastMul_helpers (Chris?)
- [x] two assume(false) in AES128.vaf and AES256.vaf (Chris?)
- [x] numerous occurrences of --admit_smt_queries true in X64.GCMencrypt.vaf *and* X64.GCMdecrypt.fst (Chris?)
- [x] avx + avx2 CPUID detection (Aymeric)
- [x] modifies_none on cpuid functions to remove an assume in EverCrypt.AutoConfig2 (Jonathan)
- [x] multiplex Poly1305 + drop old call + fix callers (Jonathan)

**For alpha1:**

- [x] `LowStar.Endianness`, a library to collect all your endianness-related needs (https://github.com/FStarLang/FStar/pull/1664) -- please go ahead and push on my branch!
  - [x] monotonic stateful operators (Aseem)
  - [x] fix `n_to_le` to take `nat` instead of `uint32` (Jonathan)
  - [x] interfaces (Jonathan)
  - [x] kremlin fixes (Jonathan)
  - [x] ~~reveal / opaque (Santiago)~~ (not doing; let's add lemmas instead)
  - [x] one leftover assume (Tahina?) from a proof that regressed
- [x] bringing poly1305, verified libraries from _dev and chacha20 in the same scope
  - [x] backport improved proofs from poly1305 in _dev_combinators into sn_evercrypt_integration (Marina)
  - [x] backport updated libraries from _dev (also in _dev_combinators) (Marina or Santiago)
  - [x] cherry-pick chacha20 from protz_failed_integration and enable it in sn_evercrypt_integration (Marina or Santiago); note: pick files in code/chacha20, specs/Spec.Chacha20.fst, specs/test/Spec.Test.Chacha20.fst, then it suffices to add chacha20 in Makefile.common https://github.com/project-everest/hacl-star/blob/sn_evercrypt_integration/Makefile.common#L22
  - [x] fix chacha20 proof regressions
  - [x] once verification is going through, fix Makefiles, linking, distribution, packaging (Jonathan)
  - [x] once this is done, rewire EverCrypt to use the new chacha20, with the understanding that the old one will remain for old-chacha-poly purposes until Aymeric comes in (see below) (Jonathan)
- [x] attempt AEAD-ChachaPoly (Aymeric)
- [x] poly1305 spec equivalences
  - [x] Vale <-> Hacl spec equivalence proof (Chris) https://github.com/project-everest/hacl-star/blob/sn_evercrypt_integration/providers/evercrypt/fst/EverCrypt.Poly1305.fst#L30
- [x] more EverCrypt tests (Jonathan)
  - [x] test vectors for Poly1305 (Jonathan)
  - [x] test vectors for Curve25519 (Jonathan)
  - [x] test vectors for ChachaPoly (Jonathan)
  - [x] test vectors for Aes128Gcm (Jonathan)
  - [x] can someone run the tests on a Goldmont machine to at least confirm that the SHA-EXT implementation works fine? **update: they don't**
- [x] AES-GCM/Vale
  - [x] finish proof (Bryan)
  - [x] Low* (?) wrapper for non-block-multiple data (?)
  - [x] interop wrapper (Aymeric)
- [x] AEAD (Jonathan)
  - [x] integrate AES-GCM from Chris with his spec (leaving off the HACL* spec for now)
  - [x] an agile AEAD operational spec for the time being (we'll leave PRF-MAC higher-order specs to later)
  - [x] fixup `#ifdef` compilation schemes (done in kremlin)
  - [x] implement `#ifdef`-based code specialization
  - [x] monotonic invariant for multiplexing underneath the expanded key
  - [x] make encrypt / decrypt call into HACL*
  - [x] make encrypt / decrypt call into Vale
  - [x] rewire legacy EverCrypt.fst to call (unverified) into new AEAD ChachaPoly (possibly via a dirty assumed symbol in EverCrypt.Hacl)
  - [x] kill code/old/chacha20, code/old/poly1305, code/old/api and remove C files from the build
  - [x] remove _new suffix for Hacl_Chacha20Poly1305
  - [x] change API to take a separate buffer for the tag (high priority)
- [x] Curve25519 (Jonathan)
  - [x] finish proof (Marina)
  - [x] integrate the development as it currently stands (Jonathan)
  - [x] kremlin c89/ifdef bug (Jonathan)
  - [x] propagate `has_abm /\ has_bmi2` throughout the HACL* development instead of assuming it (Aymeric)
  - [x] multiplex, CPU detection, proper usage of ifdefs (Jonathan)
  - [x] complete missing proofs in `Hacl.Impl.Curve25519.Field64.Core.fst` (Aymeric)
  - [x] ~~improve robustness of proofs: `FInv` seems to take a substantial amount of time; a regression in `Hacl.Spec.Curve25519.Field51.Definition.fst` (Marina)~~ (moved to: general proof robustness improvement)
  - [x] in generated inline assembly: add #ifndef FOOBAR_H #define FOOBAR_H etc. + #include <inttypes.h> (and remove Makefile hack) (Aymeric)
  - [x] in generated inline assembly: remove const (and remove Makefile hack) (Aymeric)
  - [x] in generated inline assembly: error: initialization of ‘uint64_t *’ {aka ‘long unsigned int *’} from ‘uint64_t’ {aka ‘long unsigned int’} makes pointer from integer without a cast (Aymeric)
  - [x] in generated inline assembly: curve25519-inline.h:28:1: error: no return statement in function returning non-void (Aymeric)
  - [x] fix symbol collisions (Jonathan)
  - [x] tests for curve25519 (direct HACL*) (Jonathan)
- [x] two `assume` leftover in `EverCrypt.Hash` https://github.com/project-everest/hacl-star/blob/sn_evercrypt_integration/providers/evercrypt/fst/EverCrypt.Hash.fst#L139 (Tahina)
- [x] two `assume` leftover in `X64.BufferViewStore` https://github.com/project-everest/hacl-star/blob/sn_evercrypt_integration/vale/code/arch/x64/X64.BufferViewStore.fst#L48
- [x] Improve the stack model (Aymeric)
- [x] cleanup READMEs and landing pages (Jonathan)
  - [x] terminology ("provider")
  - [x] up-to-date lists of algorithms
  - [x] fixup install instructions + docker container
- [x] https://project-everest.github.io/ to be updated to mention EverCrypt (Bryan)
- [x] merge sn_evercrypt_integration (Jonathan)
  - [x] fix src/windows/evercrypt/makefile.vs (again! can we auto-generate it?)
  - [x] #ifdef __GNUC__ in inline asm
  - [x] open_out_binary to avoid host endlines in inline asm (and regular asm too?) (now FStarLang/FStar#1683)
- [x] Assume left in X64.AESGCM.vaf (Bryan)
- [x] Stabilize proof in thirdPartyPorts/X64.AESGCM.vaf, and remove assume false (Bryan)
- [x] performance regression in Curve (Karthik)
- [x] Merkle Tree proofs
  - [x] serialization (Aseem)
- [x] add `@type: true` fsdoc annotations for the generated C code to have types (Jonathan)
- [x] fix windows assembly error (Jonathan)

**For alpha2:**

See https://github.com/project-everest/hacl-star/issues/176

**Comment by s-zanella** [comment]
I removed all remaining assumptions from `lib` in the `_dev` branch and fixed a few regressions.
`Lib.IntVector.fst` is still broken but it seems to be in a better state in `_dev_combinators`.

I believe the 3 assumes in `Hacl.Test.SHA3.fst` are harmless. The test vectors are written as immutable buffers but the SHA3 API takes mutable buffers, so the harness allocates a fresh buffer and copies over the contents of the immutable buffer before calling into SHA3. The test harness can be rewritten to avoid the `assume`, but this is low priority.

Besides completing the implementation of `Lib.IntTypes` using `FStar.ConstantTime.Integers` in the `santiago_inttypes` branch I can help integrating SHA3 into EverCrypt and verifying the `map_blocks` combinators for Chacha20.

**Comment by msprotz** [comment]
That would be wonderful. Can I let you be the libraries wizard and handle:
- the "merge" of their version in _dev into either fstar-master or protz_integration2, and
- hooking them up to LowStar.Endianness once we land it?

Thanks!

**Comment by s-zanella** [comment]
Will do.

**Comment by msprotz** [comment]
Closing in favor of https://github.com/project-everest/hacl-star/issues/176
